File: app/OnyxPenDemo/src/main/java/com/onyx/android/eink/pen/demo/erase/util/EraserTrackHelper.java
package com.onyx.android.eink.pen.demo.erase.util;

import com.onyx.android.eink.pen.demo.PenBundle;
import com.onyx.android.eink.pen.demo.erase.data.EraseTypes;
import com.onyx.android.sdk.pen.TouchHelper;
import com.onyx.android.sdk.pen.style.StrokeStyle;
import com.onyx.android.sdk.utils.ReflectUtil;

import java.lang.reflect.Method;

public final class EraserTrackHelper {

    public enum Layer {
        SF,
        APP,
        NONE
    }

    private static final String VIEW_UPDATE_HELPER_CLASS = "android.onyx.ViewUpdateHelper";
    private static final String METHOD_SET_ERASER_RAW_DRAWING_ENABLED = "setEraserRawDrawingEnabled";

    private static volatile Boolean supportsMoveStrokeSfTrack;

    private EraserTrackHelper() {
    }

    public static boolean supportsMoveStrokeSfTrack() {
        Boolean cached = supportsMoveStrokeSfTrack;
        if (cached != null) {
            return cached;
        }
        synchronized (EraserTrackHelper.class) {
            cached = supportsMoveStrokeSfTrack;
            if (cached != null) {
                return cached;
            }
            cached = probeSupportsMoveStrokeSfTrack();
            supportsMoveStrokeSfTrack = cached;
            return cached;
        }
    }

    private static boolean probeSupportsMoveStrokeSfTrack() {
        Class<?> viewUpdateHelperClass = ReflectUtil.classForName(VIEW_UPDATE_HELPER_CLASS);
        if (viewUpdateHelperClass == null) {
            return false;
        }
        Method method = ReflectUtil.getMethodSafely(
                viewUpdateHelperClass,
                METHOD_SET_ERASER_RAW_DRAWING_ENABLED,
                boolean.class,
                int.class);
        return method != null;
    }

    public static boolean defaultTrackEnabled(int eraseType) {
        switch (eraseType) {
            case EraseTypes.ERASER_AREA:
                return true;
            case EraseTypes.ERASER_MOVE:
            case EraseTypes.ERASER_STROKE:
                return supportsMoveStrokeSfTrack();
            default:
                return false;
        }
    }

    public static Layer resolveLayer(PenBundle bundle, int eraseType) {
        if (!bundle.isDisplayEraseTrack(eraseType)) {
            return Layer.NONE;
        }
        if (eraseType == EraseTypes.ERASER_AREA) {
            return Layer.SF;
        }
        if (EraseTypes.isMoveOrStrokeErase(eraseType)) {
            return supportsMoveStrokeSfTrack() ? Layer.SF : Layer.APP;
        }
        return Layer.NONE;
    }

    public static boolean useSfTrack(PenBundle bundle, int eraseType) {
        return resolveLayer(bundle, eraseType) == Layer.SF;
    }

    public static boolean useAppTrack(PenBundle bundle, int eraseType) {
        return resolveLayer(bundle, eraseType) == Layer.APP;
    }

    public static boolean shouldForceRawDrawing(PenBundle bundle, int eraseType) {
        return useSfTrack(bundle, eraseType);
    }

    public static int eraserStrokeStyle(int eraseType) {
        if (EraseTypes.isMoveOrStrokeErase(eraseType)) {
            return StrokeStyle.SOFT_ERASER;
        }
        return TouchHelper.STROKE_STYLE_DASH;
    }
}

File: app/OnyxPenDemo/src/main/java/com/onyx/android/eink/pen/demo/erase/util/EraserTrackHelper.kt
package com.onyx.android.eink.pen.demo.erase.util

import com.onyx.android.eink.pen.demo.PenBundle
import com.onyx.android.eink.pen.demo.erase.data.EraseTypes
import com.onyx.android.sdk.pen.TouchHelper
import com.onyx.android.sdk.pen.style.StrokeStyle
import com.onyx.android.sdk.utils.ReflectUtil
import kotlin.concurrent.Volatile

object EraserTrackHelper {
    private const val VIEW_UPDATE_HELPER_CLASS = "android.onyx.ViewUpdateHelper"
    private const val METHOD_SET_ERASER_RAW_DRAWING_ENABLED = "setEraserRawDrawingEnabled"

    @Volatile
    private var supportsMoveStrokeSfTrack: Boolean? = null

    fun supportsMoveStrokeSfTrack(): Boolean {
        var cached = supportsMoveStrokeSfTrack
        if (cached != null) {
            return cached
        }
        synchronized(EraserTrackHelper::class.java) {
            cached = supportsMoveStrokeSfTrack
            if (cached != null) {
                return cached
            }
            cached = probeSupportsMoveStrokeSfTrack()
            supportsMoveStrokeSfTrack = cached
            return cached
        }
    }

    private fun probeSupportsMoveStrokeSfTrack(): Boolean {
        val viewUpdateHelperClass =
            ReflectUtil.classForName(VIEW_UPDATE_HELPER_CLASS) ?: return false
        val method = ReflectUtil.getMethodSafely(
            viewUpdateHelperClass,
            METHOD_SET_ERASER_RAW_DRAWING_ENABLED,
            Boolean::class.javaPrimitiveType,
            Int::class.javaPrimitiveType
        )
        return method != null
    }

    fun defaultTrackEnabled(eraseType: Int): Boolean {
        return when (eraseType) {
            EraseTypes.ERASER_AREA -> true
            EraseTypes.ERASER_MOVE, EraseTypes.ERASER_STROKE -> supportsMoveStrokeSfTrack()
            else -> false
        }
    }

    fun resolveLayer(bundle: PenBundle, eraseType: Int): Layer {
        if (!bundle.isDisplayEraseTrack(eraseType)) {
            return Layer.NONE
        }
        if (eraseType == EraseTypes.ERASER_AREA) {
            return Layer.SF
        }
        if (EraseTypes.isMoveOrStrokeErase(eraseType)) {
            return if (supportsMoveStrokeSfTrack()) Layer.SF else Layer.APP
        }
        return Layer.NONE
    }

    fun useSfTrack(bundle: PenBundle, eraseType: Int): Boolean {
        return resolveLayer(bundle, eraseType) == Layer.SF
    }

    fun useAppTrack(bundle: PenBundle, eraseType: Int): Boolean {
        return resolveLayer(bundle, eraseType) == Layer.APP
    }

    fun shouldForceRawDrawing(bundle: PenBundle, eraseType: Int): Boolean {
        return useSfTrack(bundle, eraseType)
    }

    fun eraserStrokeStyle(eraseType: Int): Int {
        if (EraseTypes.isMoveOrStrokeErase(eraseType)) {
            return StrokeStyle.SOFT_ERASER
        }
        return TouchHelper.STROKE_STYLE_DASH
    }

    enum class Layer {
        SF, APP, NONE
    }
}
